nova_vm/src/ecmascript/builtins/global_object.rs

```diff
@@ -725,15 +728,119 @@ impl GlobalObject {
+        let s = to_string(agent, gc.reborrow(), string)?;
+        let s = s.as_str(agent).to_owned();
+
+        // 2. Let S be ! TrimString(inputString, start).
+        let s = s.trim_start_matches(is_trimmable_whitespace);
+
+        // 3. Let sign be 1.
+        // 4. If S is not empty and the first code unit of S is the code unit 0x002D (HYPHEN-MINUS), set sign to -1.
+        // 5. If S is not empty and the first code unit of S is either the code unit 0x002B (PLUS SIGN) or the code unit 0x002D (HYPHEN-MINUS), set S to the substring of S from index 1.
+        let (sign, mut s) = if let Some(s) = s.strip_prefix('-') {
+            (-1.0, s)
+        } else if let Some(s) = s.strip_prefix('+') {
+            (1.0, s)
+        } else {
+            (1.0, s)
+        };
+
+        // 6. Let R be ℝ(? ToInt32(radix)).
+        let r = to_int32(agent, gc.reborrow(), radix)?;
+
+        // 7. Let stripPrefix be true.
+        // 8. If R ≠ 0, then
+        let (mut r, strip_prefix) = if r != 0 {
+            // a. If R < 2 or R > 36, return NaN.
+            if !(2..=36).contains(&r) {
+                return Ok(Value::nan());
+            }
+            // b. If R ≠ 16, set stripPrefix to false.
+            (r as u32, r == 16)
+        } else {
+            // 9. Else,
+            // a. Set R to 10.
+            (10, true)
+        };
+
+        // 10. If stripPrefix is true, then
+        if strip_prefix {
+            // a. If the length of S is at least 2 and the first two code units of S are either "0x" or "0X", then
+            if s.starts_with("0x") || s.starts_with("0X") {
+                // i. Set S to the substring of S from index 2.
+                s = &s[2..];
+                // ii. Set R to 16.
+                r = 16;
+            }
+        };
+
+        // 11. If S contains a code unit that is not a radix-R digit, let end be the index within S of the first such code unit; otherwise, let end be the length of S.
+        let end = s.find(|c: char| !c.is_digit(r)).unwrap_or(s.len());
+
+        // 12. Let Z be the substring of S from 0 to end.
+        let z = &s[..end];
+
+        // 13. If Z is empty, return NaN.
+        if z.is_empty() {
+            return Ok(Value::nan());
+        }
+
+        // 14. Let mathInt be the integer value that is represented by Z in
+        //     radix-R notation, using the letters A through Z and a through z
+        //     for digits with values 10 through 35. (However, if R = 10 and Z
+        //     contains more than 20 significant digits, every significant
+        //     digit after the 20th may be replaced by a 0 digit, at the option
+        //     of the implementation; and if R is not one of 2, 4, 8, 10, 16,
+        //     or 32, then mathInt may be an implementation-approximated
+        //     integer representing the integer value denoted by Z in radix-R
+        //     notation.)
+        let math_int = i128::from_str_radix(z, r).unwrap() as f64;
```

Comment:
**suggestion:** We should try to avoid the f64 conversion if at all we can. Some `try_from` shenanigans would probably be a good idea here to try and avoid that. Also, for `r = 10` and `r = 2` various optimisations based on the length of `z` may be beneficial.

eg. `r = 10` if `z` length is less than 16 then we're guaranteed a safe integer value. If less than 3 then it's a guaranteed u8. If less than 5 then u16. So on. Same sort of logic can be used for `r = 2`.

Reply:
Implemented these fun fast-paths for most known safe integer values, and a macro for easily adding more if the need arises for any common radixes.